packages/react-accounts/react-accounts.ts

```diff
@@ -0,0 +1,29 @@
+import { Meteor } from 'meteor/meteor'
+import { Tracker } from 'meteor/tracker'
+import { useState, useEffect } from 'react'
+
+export const useUserId = () => {
```

Comment:
I just watched your talk in Impact Meteor 2020 and when you first introduced this guy I thought it would be global managed. I'm wondering that we are opening a computation every time we need userId in a component. I thought [zustand](https://github.com/pmndrs/zustand) would better fit this problem, although it might not be necessary because I don't know the impacts of openings many computations. What do you think? It could be applied in others states as well.

Reply:
Computations are incredibly cheap in practice. In Blaze they are the backbone of the entire system. In Meteor, for years when using `withTracker` it would actively create and destroy a computation on every react render. It's basically adding one entry to an array internally, for the reactive source. Very cheap.

Still, to recycle some of the more expensive reactive sources (not the computation itself) like an expensive reactive subscription, you can always hoist the computation up your react tree (er, or down to root), and then use the context API to feed the result of that one computation to mulitple hooks. I showed how to do that in the [hook introduction blog post](https://blog.meteor.com/introducing-usetracker-react-hooks-for-meteor-cb00c16d6222) ([here is the gist](https://gist.github.com/CaptainN/4cf649aa0a92d1ac206773495995ed47#file-meteor-hooks-kitchen-sink-jsx)). Be aware though, this can introduce some subtle timing issues with reactivity and the React lifecycle. It's easier (and cheap) to just use a lot of computations. ;-)

Reply:
One point I mean to stress whenever we ship these, is they are not solving a problem with the current HOC/hook - `useTracker` is fast as is, and we'll still recommend its use with `Collection.findOne` for example. We/I simply couldn't come up with anything better for that.

These new hooks are the result of some exploratory effort, to see if we can gain some advantage by implementing things in a more custom way. What I found was that some of my earlier assumptions (which I did express in the Impact Meteor 2020 preso) did not pan out - just exposing a cursor was actually quite a bit slower and less efficient than `useTracker`...

The main advancements we did achieve here are in 2 very specific cases (so far anyway, there may be others when we/I get to them). `useSubscription` is more optimized for subscriptions in particular. In practice, it probably won't amount to much in terms of gains, but it has a cleaner API than `useTracker` for subscriptions in particular. The second one is the big win - `useFind` is now immutable, and doesn't require the subscription to be mixed in with the cursor. This has the potential to open up new use cases for high performance reactivity in lists over what `useTracker` and `withTracker` could deliver. I say "has the potential" because these statements have not been well tested, and there are likely tradeoffs here (as with any performance optimizations). But really, the main advantage of these 2 hooks is mostly that they (arguably) improve developer experience, through a more tailored/integrated feeling API between Meteor and React, and I always love that type of improvement.

Along the way, I also discovered some improvements we should make to `useTracker` which have been submitted as a separate PR.